packages/api-codec/src/base/Array.ts

```diff
@@ -0,0 +1,117 @@
+// Copyright 2017-2018 @polkadot/api-codec authors & contributors
+// This software may be modified and distributed under the terms
+// of the ISC license. See the LICENSE file for details.
+
+import u8aConcat from '@polkadot/util/u8a/concat';
+
+import CodecBase from './Base';
+import Length from './LengthCompact';
+
+// This manages codec arrays. Intrernally it keeps track of the length (as decoded) and allows
+// construction with the passed `Type` in the constructor. It aims to be an array-like structure,
+// i.e. while it wraps an array, it provides a `length` property to get the actual length, `at(index)`
+// to retrieve a specific item. Additionally the helper functions `map`, `filter`, `forEach` and
+// `reduce` is exposed on the interface.
+export default class CodecArray <
+  T extends CodecBase
+> extends CodecBase<Array<T>> {
+  private _length: Length;
+  private _Type: { new(value?: any): T };
+
+  constructor (Type: { new(value?: any): T }, value: Array<any> = [] as Array<any>) {
+    super(
+      value.map((entry) =>
+        new Type(entry)
+      )
+    );
+
+    this._length = new Length(value.length);
+    this._Type = Type;
+  }
+
+  static with <O extends CodecBase> (Type: { new(value?: any): O }): { new(value?: any): CodecArray<O> } {
+    return class extends CodecArray<O> {
+      constructor (value?: Array<any>) {
+        super(Type, value);
+      }
+    };
+  }
+
+  byteLength (): number {
+    return this.raw.reduce((total, raw) => {
+      return total + raw.byteLength();
+    }, this._length.byteLength());
+  }
+
+  get length (): number {
+    return this.raw.length;
+  }
+
+  at (index: number): T {
+    return this.raw[index];
+  }
+
+  filter (fn: (entry: T, index?: number) => any): Array<T> {
+    return this.raw.filter(fn);
+  }
+
+  forEach (fn: (entry: T, index?: number) => any): any {
+    return this.raw.forEach(fn);
+  }
+
+  fromJSON (input: any): CodecArray<T> {
```

Comment:
Agreed actually. However the reason I went down this route is twofold -

1. For the stuff where we need to know the actual type, it is probably easier constructing with the class and then where we from it can be used.

2. ... erm, I started with it that way around. However TS basically doesn't allow statics in interfaces (where I started) and it became one big hack to try and work around this.

Not 100% crazy about the above, happy to see it changed.